physical/raft/fsm.go

```diff
@@ -243,6 +252,87 @@ func writeSnapshotMetaToDB(metadata *raft.SnapshotMeta, db *bolt.DB) error {
 	return nil
 }
 
+func (f *FSM) upgradeDesiredSuffrage() error {
+	// Read the desired suffrage entry
+	var dsBytes []byte
+	if err := f.db.View(func(tx *bolt.Tx) error {
+		value := tx.Bucket(configBucketName).Get(desiredSuffrageKey)
+		if value != nil {
+			dsBytes = make([]byte, len(value))
+			copy(dsBytes, value)
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	var ds DesiredSuffrageValue
+
+	// Entry is already present. Read and get the suffrage value.
+	if dsBytes != nil {
+		err := proto.Unmarshal(dsBytes, &ds)
+		if err != nil {
+			return err
+		}
+		f.desiredSuffrage = ds.DesiredSuffrage
+		return nil
+	}
+
+	//
+	// This is the upgrade case where there is no entry
+	//
+
+	// Refer to the persisted latest raft config
+	config := f.latestConfig.Load().(*ConfigurationValue)
+
+	// If there is no config, then this is a fresh node coming up. This could end up
+	// being a voter or non-voter. But by default assume that this is a voter. It
+	// will be changed if this node joins the cluster as a non-voter.
+	if config == nil {
+		ds.DesiredSuffrage = f.desiredSuffrage
+		return f.persistDesiredSuffrage(&ds)
+	}
+
+	// Get the last known suffrage of the node and assume that it is the desired
+	// suffrage. There is no better alternative here.
+	for _, srv := range config.Servers {
+		if srv.Id == f.localID {
+			switch srv.Suffrage {
+			case int32(raft.Nonvoter):
+				ds.DesiredSuffrage = "non-voter"
+			default:
+				ds.DesiredSuffrage = "voter"
+			}
+		}
+		// Bring the intent to the fsm instance.
+		f.desiredSuffrage = ds.DesiredSuffrage
+	}
+
+	return f.persistDesiredSuffrage(&ds)
+}
+
+// witnessSuffrage is called when a node successfully joins the cluster. This
+// intent should land in the stored configuration. If the config isn't available
+// yet, we still go ahead and store the intent in the fsm. During the next
+// update to the configuration, this intent will be persisted.
+func (f *FSM) witnessSuffrage(desiredSuffrage string) error {
```

Comment:
"witness" doesn't evoke the right thing to me for what this does - how about "recordSuffrage"?

Reply:
Yep. Sounds good. Done.